app/src/main/java/com/mikhailovskii/weatherandroid/data/firebase/FirebaseModel.kt

```diff
@@ -0,0 +1,188 @@
+
+            for (document in result) {
+                val stickerPack = document.toObject(StickerPack::class.java)
+                stickerPackList.add(stickerPack)
+            }
+
+            callback?.onFirebaseDataLoaded(stickerPackList)
+        }.addOnFailureListener {
+            callback?.onFirebaseDataFailed()
+        }
+
+    }
+
+    fun saveUserLocation(user: User?) {
+        db.collection(USERS_COLLECTION).whereEqualTo(LOGIN_FIELD, user?.login).get()
+            .addOnSuccessListener { databaseResult ->
+
+                val map: Map<String, String> = hashMapOf(LOCATION to (user?.location ?: ""))
+
+                for (document in databaseResult) {
+                    db.collection(USERS_COLLECTION).document(document.id).update(map)
+                }
+
+            }
+    }
+
+    fun baseLogInUser(login: String, password: String, callback: FirebaseLoginCallback?) {
+        db.collection(USERS_COLLECTION).get().addOnSuccessListener { result ->
+            for (document in result) {
+                var user = document.toObject(User::class.java)
+
+                if (user.login == login && user.password == password) {
+                    user = User(login = login, password = password)
+
+                    if (user.location?.isNotBlank() == true) {
+                        callback?.onLoggedInWithFilledInfo()
+                    } else {
+                        callback?.onLoggedInWithEmptyLocation()
+                    }
+                }
+            }
+        }.addOnFailureListener {
+            callback?.onLoginFailed()
+        }
+    }
+
+    fun logInWithTwitter(
+        result: Result<com.twitter.sdk.android.core.models.User>?,
+        user: User,
+        callback: FirebaseLoginCallback?
+    ) {
+        db.collection(USERS_COLLECTION).get().addOnSuccessListener { databaseResult ->
+            var isUserPresents = false
+
+            loop@ for (document in databaseResult) {
+                val databaseUser = document.toObject(User::class.java)
+
+                if (databaseUser.login == result?.data?.name
+                    && databaseUser.twitterKey == result?.data?.idStr
+                    && databaseUser.icon == result?.data?.profileImageUrl
+                ) {
+                    isUserPresents = true
+                    break@loop
+                }
+            }
+
+            if (!isUserPresents) {
+                db.collection(USERS_COLLECTION).document().set(user)
+                    .addOnSuccessListener {
+                        Timber.d("Twitter info saved")
+                    }.addOnFailureListener { e ->
+                        Timber.e("Error twitter save: $e")
+                    }
+            }
+
+            callback?.onLoggedInWithFilledInfo()
+        }.addOnFailureListener {
+            callback?.onLoginFailed()
+        }
+
+    }
+
+    fun logInWithFacebook(user: User, callback: FirebaseLoginCallback?) {
```

Comment:
@SergeiMikhailovskii think how to unify logic from `logInWithTwitter`, `logInWithFacebook` and `logInWithGoogle`. It's 90% the same now.